asciimatics/widgets/frame.py

```diff
@@ -15,6 +15,85 @@
 from asciimatics.widgets.utilities import THEMES, logger
 
 
+class _BorderManager:
+    def __init__(self, frame, has_border, can_scroll):
+        self._frame = frame
+        self.has_border = has_border
+        self._scroll_bar = None
+        if can_scroll:
+            scroll_y = 2
+            scroll_height = frame._canvas.height - 4
+            if not has_border:
+                scroll_height = frame._canvas.height - 2
+                scroll_y = 1
+
+            self._scroll_bar = _ScrollBar(frame._canvas, frame.palette, 
+                frame._canvas.width - 1, scroll_y, scroll_height,
+                frame._get_pos, frame._set_pos, absolute=True) 
+
+        self.tl    = u"┌" if frame._canvas.unicode_aware else "+"
+        self.tr    = u"┐" if frame._canvas.unicode_aware else "+"
+        self.bl    = u"└" if frame._canvas.unicode_aware else "+"
+        self.br    = u"┘" if frame._canvas.unicode_aware else "+"
+        self.horiz = u"─" if frame._canvas.unicode_aware else "-"
+        self.vert  = u"│" if frame._canvas.unicode_aware else "|"
+
+        # Optimization for non-unicode displays to avoid slow unicode calls.
+        self.string_len = wcswidth if frame._canvas.unicode_aware else len
+
+    @property
+    def can_scroll(self):
+        return self._scroll_bar is not None
+
+    def get_rectangle(self):
```

Comment:
Definitely worth an explanation of what reactangle this is here...  IIUC, it's the usable rectangle left in the Canvas taking into account any borders or scroll bars.